drivers/charger/charger_bq25180.c

```diff
@@ -134,10 +140,68 @@ static int bq25180_get_charge_current(const struct device *dev,
 	return 0;
 }
 
+static int bq25180_get_online(const struct device *dev,
+			      enum charger_online *online)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & BQ25180_STAT0_VIN_PGOOD_STAT) != 0x00) {
+		*online = CHARGER_ONLINE_FIXED;
+	} else {
+		*online = CHARGER_ONLINE_OFFLINE;
+	}
+
+	return 0;
+}
+
+static int bq25180_get_status(const struct device *dev,
+			      enum charger_status *status)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & BQ25180_STAT0_VIN_PGOOD_STAT) == 0x00) {
+		*status = CHARGER_STATUS_DISCHARGING;
+		return 0;
+	}
+
+	switch (FIELD_GET(BQ25180_STAT0_CHG_STAT_MASK, val)) {
+	case BQ25180_STAT0_CHG_STAT_NOT_CHARGING:
+		*status = CHARGER_STATUS_NOT_CHARGING;
+		break;
+	case BQ25180_STAT0_CHG_STAT_CONSTANT_CURRENT:
+	case BQ25180_STAT0_CHG_STAT_CONSTANT_VOLTAGE:
+		*status = CHARGER_STATUS_CHARGING;
+		break;
+	case BQ25180_STAT0_CHG_STAT_DONE:
+		*status = CHARGER_STATUS_FULL;
```

Comment:
So some TI chargers have this ambiguity that is mentioned in table 8-9 of the datasheet:

> 2b11 = Charge Done or charging is disabled by the host.

The charge cycle could be completed or the `CHG_DIS` could be 1. To avoid this ambiguity, I suggest checking `CHG_DIS` before this switch and after you check `PGOOD_STAT` like so

``` c
ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, val);
if (ret < 0) {
	return ret;
}

if (val & BQ25180_ICHG_CHG_DIS) {
	*status = CHARGER_STATUS_NOT_CHARGING;
	return 0;
}
```
Sorry it's kind of annoying that you'll either need to cache both registers or reread `BQ25180_STAT0`

Reply:
Thanks, I did stare at that table and the `charger_status` values for a bit and tried to figure how to map it, this makes sense.

I don't think the multiple transaction is a big deal, doubt this gets polled often.